File: lib/models/urine_report.dart
class UrineReport {
  final int id;
  final int userId;
  final DateTime testDate;
  final String color;
  final String appearance;
  final String protein;
  final String sugar;
  final double specificGravity;
  final String? imageUrl;

  UrineReport({
    required this.id,
    required this.userId,
    required this.testDate,
    required this.color,
    required this.appearance,
    required this.protein,
    required this.sugar,
    required this.specificGravity,
    this.imageUrl,
  });

  // Factory constructor to create a UrineReport instance from a map (JSON)
  factory UrineReport.fromJson(Map<String, dynamic> json) {
    return UrineReport(
      id: json['id'] as int,
      userId: json['user_id'] as int,
      testDate: DateTime.parse(json['test_date'] as String),
      color: json['color'] as String,
      appearance: json['appearance'] as String,
      protein: json['protein'] as String,
      sugar: json['sugar'] as String,
      specificGravity: json['specific_gravity'] as double,
      imageUrl: json['image_url'] as String?,
    );
  }

  // Method to convert a UrineReport instance to a map (JSON)
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'test_date': testDate.toIso8601String(),
      'color': color,
      'appearance': appearance,
      'protein': protein,
      'sugar': sugar,
      'specific_gravity': specificGravity,
      'image_url': imageUrl,
    };
  }
}

File: models-kotlin/UrineReport.kt
package com.healthtracker.models

import com.google.gson.annotations.SerializedName
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

data class UrineReport(
    val id: Int,
    @SerializedName("user_id")
    val userId: Int,
    @SerializedName("test_date")
    val testDate: LocalDateTime,
    val color: String,
    val appearance: String,
    val protein: String,
    val sugar: String,
    @SerializedName("specific_gravity")
    val specificGravity: Double,
    @SerializedName("image_url")
    val imageUrl: String? = null
) {
    companion object {
        private val dateFormatter = DateTimeFormatter.ISO_LOCAL_DATE_TIME

        fun fromJson(json: Map<String, Any?>): UrineReport {
            return UrineReport(
                id = json["id"] as Int,
                userId = json["user_id"] as Int,
                testDate = LocalDateTime.parse(json["test_date"] as String, dateFormatter),
                color = json["color"] as String,
                appearance = json["appearance"] as String,
                protein = json["protein"] as String,
                sugar = json["sugar"] as String,
                specificGravity = json["specific_gravity"] as Double,
                imageUrl = json["image_url"] as String?
            )
        }
    }

    fun toJson(): Map<String, Any?> {
        return mapOf(
            "id" to id,
            "user_id" to userId,
            "test_date" to testDate.format(dateFormatter),
            "color" to color,
            "appearance" to appearance,
            "protein" to protein,
            "sugar" to sugar,
            "specific_gravity" to specificGravity,
            "image_url" to imageUrl
        )
    }
}